### ISSUE TITLE
[Bug] Loading screen (kind of) freezes?

### ISSUE THREAD

**Issue by pjalsGit** [WerWolv/ImHex#315]
### Operating System

Linux

### What's the issue you encountered?

On Fedora 34 I tried to run ImHex, and the loading screen initiated, and dissapeared, but the program never appeared.

### How can the issue be reproduced?

1. Clone ImHex
2. Get Submodules
3. Build
4. Run builded result (on Fedora 34)
5. See bug

### ImHex Version

aef9598 master

### Additional context?

Logs:
```
terminate called after throwing an instance of 'nlohmann::detail::type_error'
  what():  [json.exception.type_error.302] type must be number, but is null
```
(thats it)

**Comment by WerWolv** [comment]
Huh I thought I fixed this issue for good. Try deleting ImHex's settings.json file that should have been created in `$HOME/.config/imhex/config`

**Comment by pjalsGit** [comment]
same error, by the way i forgot to mention, you need to use kill on it with signal KILL to stop it

**Comment by pjalsGit** [comment]
whoops!

**Comment by raron** [comment]
I think have the same issue and deleting the file `$HOME/.config/imhex/config` does not solve it. (The file is re-created with 4 byte content: `null`.)

---

**tl;dr: it seems to me the ImHex crashes in this way at startup iff the built-in plugin (buildin.hexplug) is missing AND the settings.json file is missing / empty / non-valid at the same time**  

* if the plugin can be loaded, and the settings.json file is missing, it setting file will be created with valid content
* if the plugin can NOT be loaded, and the settings.json file is valid, the ImHex will show an error message about the plugin load failure  
* if if the plugin can NOT be loaded, and the settings.json file is missing / non-valid, the ImHex crashes at startup

You can check from where tries ImHex load the builtin plugin, and if it is successful or not with strace:

    $ strace -f -e openat,stat ./imhex 2>&1 | grep "plugins"
    [pid  2243] stat("/tmp/e/plugins", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
    [pid  2243] openat(AT_FDCWD, "/tmp/e/plugins", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 22
    [pid  2243] openat(AT_FDCWD, "/tmp/e/plugins/builtin.hexplug", O_RDONLY|O_CLOEXEC) = 20

---

I have tested with both commit ID 2176988 and 4e52093.

A bit of debugging from my side:

It crashes at `ImHex/source/views/view_hexeditor.cpp:207` when in `EventManager::subscribe<EventSettingsChanged>` tries to it execute `auto alpha = ContentRegistry::Settings::getSetting(...)`

The call to the `getSetting` does not show up in the backtrace (maybe because it is in the shared lib?) but I can examine it by putting a breakpoint on it.

In the function `getSetting` at `ImHex/plugins/libimhex/source/api/content_registry.cpp:142` the call to `getSettingsData()` is ok, but the it returns with an empty object at line 145, (I suppose) because the `settings.contains(unlocalizedCategory)` returns false (with `unlocalizedCategory`: "hex.builtin.setting.interface") (which seems plausible with a deleted / empty settings file).


gdb backtrack:

    #0  0x00007ffff68897bb in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
    #1  0x00007ffff6874535 in __GI_abort () at abort.c:79
    #2  0x00007ffff6c1c911 in  () at ./libstdc++.so.6
    #3  0x00007ffff6c2838c in  () at ./libstdc++.so.6
    #4  0x00007ffff6c27369 in  () at ./libstdc++.so.6
    #5  0x00007ffff6c27d21 in __gxx_personality_v0 () at ./libstdc++.so.6
    #6  0x00007ffff6a23e33 in  () at /lib/x86_64-linux-gnu/libgcc_s.so.1
    #7  0x00007ffff6a246b5 in _Unwind_Resume () at /lib/x86_64-linux-gnu/libgcc_s.so.1
    #8  0x000055555565ce1e in operator()() const (__closure=0x7fffec001ed0) at /home/aron/a/ImHex/source/views/view_hexeditor.cpp:207
    #9  0x000055555566ea80 in std::__invoke_impl<void, hex::ViewHexEditor::ViewHexEditor()::<lambda()>&>(std::__invoke_other, struct {...} &) (__f=...) at /usr/i
    #10 0x000055555566bb37 in std::__invoke_r<void, hex::ViewHexEditor::ViewHexEditor()::<lambda()>&>(struct {...} &) (__fn=...) at /usr/include/c++/10/bits/invo
    #11 0x0000555555669295 in std::_Function_handler<void(), hex::ViewHexEditor::ViewHexEditor()::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
        at /usr/include/c++/10/bits/std_function.h:291
    #12 0x00005555555ff012 in std::function<void ()>::operator()() const (this=0x7fffec001ed0) at /usr/include/c++/10/bits/std_function.h:622
    #13 0x0000555555603302 in hex::Event<>::operator()() const (this=0x7fffec001ed0) at /home/aron/a/ImHex/plugins/libimhex/include/hex/api/event.hpp:49
    #14 0x00005555555fe893 in hex::EventManager::post<hex::EventSettingsChanged>() () at /home/aron/a/ImHex/plugins/libimhex/include/hex/api/event.hpp:89
    #15 0x00005555555f1aad in hex::Window::Window() (this=0x7fffffffdbc0) at /home/aron/a/ImHex/source/window/window.cpp:257
    #16 0x00005555555cb915 in main(int, char**) (argc=1, argv=0x7fffffffdfa8) at /home/aron/a/ImHex/source/main.cpp:37

I have build it with:  
cmake version 3.16.3  
g++ version g++-10 (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0  
on an Ubuntu 20.04.3 LTS  
(running as an LXC container, but I do not think this makes any difference).

If you want I can share a coredump, but it is more than 300 MiB.

---

Okay, after a bit more debugging, I come up some path:

    find -name '*.cpp' -print0 | xargs -0 perl -pi -e 's/(\s*)(auto\s*([a-zA-Z0-9_]*)\s*=\s*ContentRegistry::Settings::getSetting[^;]*;)/\1\2\n\1if (\3.is_null()) return;/g'

this adds some error handling after each `getSettings` call, like

             EventManager::subscribe<EventSettingsChanged>(this, [this] {
                 {
                     auto alpha = ContentRegistry::Settings::getSetting("hex.builtin.setting.interface", "hex.builtin.setting.interface.highlight_alpha");
    +                if (alpha.is_null()) return;
     
                     this->m_highlightAlpha = alpha;
                 }

After this the ImHex starts up and creates a (seems to be) valid setting.json file:

    $ cat ./.config/imhex/settings.json | jq 
    {
      "hex.builtin.setting.general": {
        "hex.builtin.setting.general.auto_load_patterns": 1,
        "hex.builtin.setting.general.show_tips": 1
      },
      "hex.builtin.setting.hex_editor": {
        "hex.builtin.setting.hex_editor.advanced_decoding": 0,
        "hex.builtin.setting.hex_editor.ascii": 1,
        "hex.builtin.setting.hex_editor.column_count": 16,
        "hex.builtin.setting.hex_editor.extra_info": 1,
        "hex.builtin.setting.hex_editor.grey_zeros": 1,
        "hex.builtin.setting.hex_editor.hexii": 0,
        "hex.builtin.setting.hex_editor.uppercase_hex": 1
      },
      "hex.builtin.setting.interface": {
        "hex.builtin.setting.interface.color": 0,
        "hex.builtin.setting.interface.fps": 60,
        "hex.builtin.setting.interface.highlight_alpha": 128,
        "hex.builtin.setting.interface.language": "en-US",
        "hex.builtin.setting.interface.scaling": 0
      }
    }

After the settings.json file is created the non-pathed version of ImHex starts up, too.

---

Hmm it printed an error message about it can not load even the built-in plugin, I have put the buildin.hexplug to the correct location, and now ImHex works, even if the setting.json file is deleted.

**Comment by raron** [comment]
> i got somewhere, but now it isnt loading localization, like, File menu becomes hex.menu.file, etc

I suspect you it can not load the "built-in plugin".

Try to find the plugins directory: eg. `./plugins`, or `~/.local/share/imhex/plugins` or `/usr/share/imhex/plugins` or `/usr/local/share/imhex/plugins`. There should be a file named `builtin.hexplug` in one of them.

You can find the `builtin.hexplug` in the directory `build_directory/plugins/builtin/` if you built it yourself.

(Please note that I am not a contributor of this project, I only had the same / similar problem and dig a bit.)

**Comment by pjalsGit** [comment]
strangely enough, it is loading the plugin

**Comment by pjalsGit** [comment]
since im too lazy to build, and theres new commits, has this been fixed?

**Comment by DaQueenJodi** [comment]
I have just built it with the latest commit and am facing the same issue. even completely deleting `plugins/builtin/` deleting `~/.config/imhex` doesn't resolve the issue. 

for context, I'm on archlinux and using homebrew to install mbedtls if that matters (archlinux's mbedtls doesnt include the proper .so files for imhex, so I had to use homebrew)

**Comment by WerWolv** [comment]
Fixed the crash. The thing that's causing the crash however is that the built-in plugin isn't being loaded properly. Make sure it's either in the plugins folder directly next to the application or in `/usr/share/imhex/plugins`

**Comment by dmgolembiowski** [comment]
> Fixed the crash. The thing that's causing the crash however is that the built-in plugin isn't being loaded properly. Make sure it's either in the plugins folder directly next to the application or in `/usr/share/imhex/plugins`

Hey @WerWolv , I was wondering if you could clarify what steps fix the behavior? I believe I've followed your steps as described, but I'm still experiencing the problem. Is it necessary to switch to a different branch and rebuild from source (i.e. from `rust_content_registry` to `master`) and then retry something?

**Comment by dmgolembiowski** [comment]
> > Fixed the crash. The thing that's causing the crash however is that the built-in plugin isn't being loaded properly. Make sure it's either in the plugins folder directly next to the application or in `/usr/share/imhex/plugins`
> 
> Hey @WerWolv , I was wondering if you could clarify what steps fix the behavior? I believe I've followed your steps as described, but I'm still experiencing the problem. Is it necessary to switch to a different branch and rebuild from source (i.e. from `rust_content_registry` to `master`) and then retry something?

Nevermind, for whatever reason my `~/.config/imhex/settings.json` held only `null`. After updating the file's contents with @raron's (above) it works!
